src/query_builder.rs

```diff
@@ -0,0 +1,737 @@
+    }
+
+    fn as_timestamp(mut self) -> Option<i64> {
+        self.0.take().map_or(None, |t| t.as_timestamp())
+    }
+
+    fn as_string(mut self) -> Option<Rc<String>> {
+        self.0.take().map_or(None, |t| t.as_string())
+    }
+
+    fn as_uuid(mut self) -> Option<Uuid> {
+        self.0.take().map_or(None, |t| t.as_uuid())
+    }
+
+    fn as_uuid_string(mut self) -> Option<String> {
+        self.0.take().map_or(None, |t| t.as_uuid_string())
+    }
+}
+
+pub struct CollResult(pub Vec<TypedValue>);
+
+impl CollResult {
+    fn get_row(&self, row: usize) -> Option<TypedValue> {
+        self.0.get(row).map(|v| v.to_owned())
+    }
+
+    pub fn iter(&self) -> ::std::slice::Iter<TypedValue> {
+        self.0.iter()
+    }
+
+    pub fn len(&self) -> usize {
+        self.0.len()
+    }
+}
+
+impl ColResultsRow for CollResult {
+    fn col_as_entid(&self, row: usize) -> Option<Entid> {
+        self.get_row(row).map_or(None, |t| t.as_entid())
+    }
+
+    fn col_as_kw(&self, row: usize) -> Option<Rc<NamespacedKeyword>> {
+        self.get_row(row).map_or(None, |t| t.as_kw())
+    }
+
+    fn col_as_boolean(&self, row: usize) -> Option<bool> {
+        self.get_row(row).map_or(None, |t| t.as_boolean())
+    }
+
+    fn col_as_long(&self, row: usize) -> Option<i64> {
+        self.get_row(row).map_or(None, |t| t.as_long())
+    }
+
+    fn col_as_double(&self, row: usize) -> Option<f64> {
+        self.get_row(row).map_or(None, |t| t.as_double())
+    }
+
+    fn col_as_instant(&self, row: usize) -> Option<DateTime<Utc>> {
+        self.get_row(row).map_or(None, |t| t.as_instant())
+    }
+
+    fn col_as_timestamp(&self, row: usize) -> Option<i64> {
+        self.get_row(row).map_or(None, |t| t.as_timestamp())
+    }
+
+    fn col_as_string(&self, row: usize) -> Option<Rc<String>> {
+        self.get_row(row).map_or(None, |t| t.as_string())
+    }
+
+    fn col_as_uuid(&self, row: usize) -> Option<Uuid> {
+        self.get_row(row).map_or(None, |t| t.as_uuid())
+    }
+
+    fn col_as_uuid_string(&self, row: usize) -> Option<String> {
+        self.get_row(row).map_or(None, |t| t.as_uuid_string())
+    }
+}
+
+impl IntoIterator for CollResult {
+    type Item = TypedValue;
+    type IntoIter = ::std::vec::IntoIter<TypedValue>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.0.into_iter()
+    }
+}
+
+pub struct TupleResult {
+    map: IndexMap<String, TypedValue>,
+}
+
+impl TupleResult {
+    pub fn new(spec: &Vec<String>, values: Vec<TypedValue>) -> TupleResult {
+        let mut map = IndexMap::new();
+        for (index, element) in spec.iter().enumerate() {
+            if let Some(value) = values.get(index).take() {
+                map.insert(element.clone(), value.to_owned());
+            }
+        }
+        TupleResult {map}
+    }
+
+    pub fn iter(&self) -> indexmap::map::Iter<String, TypedValue> {
+        self.map.iter()
+    }
+
+    fn get_value_for_var(&self, var: &str) -> Option<TypedValue> {
```

Comment:
Hm, interesting. This is not a good fit for aggregates. Does it 'feel' better in use than indexed lookups?